bigframes/core/transpiler/googlesql/query.py

```diff
@@ -0,0 +1,151 @@
+
+from __future__ import annotations
+
+import dataclasses
+import typing
+
+import bigframes.core.transpiler.googlesql.abc as abc
+import bigframes.core.transpiler.googlesql.expression as expr
+
+"""This module provides a structured representation of GoogleSQL syntax using nodes.
+Each node's name and child nodes are designed to strictly follow the official GoogleSQL
+syntax rules outlined in the documentation:
+https://cloud.google.com/bigquery/docs/reference/standard-sql/query-syntax"""
+
+
+@dataclasses.dataclass
+class QueryExpr(abc.SQLSyntax):
+    """This class represents GoogleSQL `query_expr` syntax."""
+
+    select: Select
+    with_cte_list: typing.Sequence[NonRecursiveCTE] = ()
+
+    def sql(self) -> str:
+        text = []
+        if len(self.with_cte_list) > 0:
+            with_cte_text = ",\n".join(
+                [with_cte.sql() for with_cte in self.with_cte_list]
+            )
+            text.append(f"WITH {with_cte_text}")
+
+        text.append(self.select.sql())
+        return "\n".join(text)
+
+
+@dataclasses.dataclass
+class Select(abc.SQLSyntax):
+    """This class represents GoogleSQL `select` syntax."""
+
+    select_list: typing.Sequence[SelectExpression]
+    from_clause_list: typing.Sequence[FromClause] = ()
+
+    def sql(self) -> str:
+        text = ["SELECT"]
+
+        select_list_sql = ",\n".join([select.sql() for select in self.select_list])
+        text.append(select_list_sql)
+
+        if self.from_clause_list is not None:
+            from_clauses_sql = ",\n".join(
+                [clause.sql() for clause in self.from_clause_list]
+            )
+            text.append(f"FROM\n{from_clauses_sql}")
+        return "\n".join(text)
+
+
+@dataclasses.dataclass
+class SelectExpression(abc.SQLSyntax):
+    """This class represents `select_expression` and `select_all` (aka. `SELECT *`)
+    syntaxes from GoogleSQL. Merges both syntaxes because `expr.Expression` can represent
+    single columns or the '*' wildcard.
+    """
+
+    expression: expr.Expression
+    alias: typing.Optional[expr.AliasExpression] = None
+
+    def __post_init__(self):
+        if isinstance(self.expression, expr.StarExpression) and self.alias is not None:
+            raise ValueError("Cannot alias when select star.")
+
+    def sql(self) -> str:
+        if self.alias is None:
+            return self.expression.sql()
+        else:
+            return f"{self.expression.sql()} AS {self.alias.sql()}"
+
+
+@dataclasses.dataclass
+class FromClause(abc.SQLSyntax):
+    """This class represents GoogleSQL `from_clause` syntax."""
+
+    from_item: FromItem
+
+    def sql(self) -> str:
+        return self.from_item.sql()
+
+
+@dataclasses.dataclass
+class FromItem(abc.SQLSyntax):
+    """This class represents GoogleSQL `from_item` syntax."""
+
+    table_name: typing.Optional[expr.TableExpression] = None
+    query_expr: typing.Optional[QueryExpr | str] = None
+    cte_name: typing.Optional[expr.Expression] = None
```

Comment:
Maybe this should be a single union field, since they are mutually exclusive

Reply:
Good idea. I tried to replace with `name: typing.Union[expr.TableExpression, QueryExpr, expr.CTEExpression]`. However, this expression does not reflect the syntax as `from_item := table_name | (query_expr) | cte_name`, though the type can represent the name. Hence, I would hold the combination for now but will revisit at b/345844377